Introduce pure python mode in language basics documentation

matusvalo opened the pull request:
This PR is introducing pure python mode in language basics documentation. It is partial fix of https://github.com/cython/cython/issues/4187

matusvalo commented:
This PR adds only the code examples in pure python form. The text is not adjusted. I am not sure whether this is ideal state since the text is focusing and referring to cython language. Does it make sence to have cython tabs default ones? I think it will fit better with the text of manual. (maybe this topic is better for discussion in #4187)

matusvalo commented:
Another point: I was not able to convert everything in the page - e.g. enums, example of pointer to function:
```
int (*grail)(int, char*) except -1
```

I have also found a bug in pure python mode not allowing convert file `examples/userguide/language_basics/parameter_refcount.py` to pure python mode

da-woods commented:
I think we're expecting that some bits aren't convertible, and that's probably ok.

It might be worth creating an issue for "language features that don't yet work in Pure Python mode" though, just to see what's missing

0dminnimda commented:
> Does it make sence to have cython tabs default ones?

We have a convention that the first tab, and therefore the default tab, is pure python.

See [here](https://github.com/cython/cython/issues/4187#issuecomment-860123077_

0dminnimda commented:
The idea is that in the future, when type annotations become an integral part of the language that will be taught to everyone then, using pure Python mode, it will be easier for beginners to start using cython.

scoder commented on docs/examples/userguide/language_basics/casting_python.py:
```diff
@@ -0,0 +1,18 @@
+
+    python_string = "foo"
+
+    ptr: p_void = cython.cast(cython.p_void, python_string)
+    adress_in_c: Py_intptr_t= cython.cast(Py_intptr_t, ptr)
+    address_from_void = adress_in_c        # address_from_void is a python int
+
+    ptr2: cython.pointer(PyObject) = cython.cast(cython.pointer(PyObject), python_string)
+    address_in_c2: Py_intptr_t= cython.cast(Py_intptr_t, ptr2)
+    address_from_PyObject = address_in_c2  # address_from_PyObject is a python int
```
I think we can avoid cluttering the example with type declarations when the type is "obviously" safe to infer.
```suggestion
    # Note that the variables below are automatically inferred
    # as the correct pointer type that is assigned to them.
    # They do not need to be typed explicitly.
    
    ptr = cython.cast(cython.p_void, python_string)
    adress_in_c = cython.cast(Py_intptr_t, ptr)
    address_from_void = adress_in_c        # address_from_void is a python int

    ptr2 = cython.cast(cython.pointer(PyObject), python_string)
    address_in_c2 = cython.cast(Py_intptr_t, ptr2)
    address_from_PyObject = address_in_c2  # address_from_PyObject is a python int
```

scoder commented on docs/examples/userguide/language_basics/open_file.py:
```diff
@@ -0,0 +1,18 @@
+from cython.cimports.cpython.exc import PyErr_SetFromErrnoWithFilenameObject
+
+def open_file():
+    p: cython.pointer(FILE) = fopen("spam.txt", "r")
```
```suggestion
    p = fopen("spam.txt", "r")   # The type of "p" is "FILE*", as returned by fopen().
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -45,61 +48,104 @@ C variable and type definitions
 ===============================
 
 The :keyword:`cdef` statement is used to declare C variables, either local or
-module-level::
+module-level
```
This statement is incorrect for the Python version. Also, since type hints are ignored for global variables, this needs to use the explicit `cython.declare()` command.

0dminnimda replied:
> this needs to use the explicit cython.declare() command.

It would be very good to indicate this, for example in a comment or explanation, since this behavior has not yet been documented, although in a good way it belongs [here](https://cython.readthedocs.io/en/latest/src/tutorial/pure.html).

0dminnimda replied:
In addition, you can make the pure python version match the version with `cdef` just with `cython.declare()`.
i.e. `cython.declare(i=cython.int, j=cython.int, k=cython.int)`

and for alignment reasons, `import cython` is not needed here, especially since it has already been written about in `two-syntax-variants-used`

matusvalo replied:
done.

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -45,61 +48,104 @@ C variable and type definitions
+                soft = 2
+                runny = 3
+
+
+
```
I don't think we need that much space here.
```suggestion
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -45,61 +48,104 @@ C variable and type definitions
 
-There is currently no special syntax for defining a constant, but you can use
-an anonymous :keyword:`enum` declaration for this purpose, for example,::
+            Structs can be declared as ``cdef packed struct``, which has
+            the same effect as the C directive ``#pragma pack(1)``.
```
Looks like there's no way to do this in Python syntax. I created #4253 to track it.

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -45,61 +48,104 @@ C variable and type definitions
+                ctypedef int* IntPtr
 
 
 It is also possible to declare functions with :keyword:`cdef`, making them c functions.
```
This is confusing when showing the Python syntax version. Probably also applies to other places.

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -135,9 +189,21 @@ whereas ``x[0]`` is.
-    cdef list foo = []
+        .. code-block:: python
+
+            foo: list = []
```
This should better be wrapped in a function to make it actually do something. (Same below.)

matusvalo replied:
done.

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -562,11 +806,25 @@ leaving ``s`` dangling. Since this code could not possibly work, Cython refuses
+
+        .. code-block:: python
+
+            s: cython.p_char
+            p = pystring1 + pystring2
+            s = p
```
This needs a function.

matusvalo replied:
done.

scoder reviewed (commented):
- The explanatory texts need the be adjusted to match both variants.
- Avoid type hints on global variables. Readers will just try them out as they are and be confused when things fail in weird ways.

matusvalo commented on docs/src/userguide/language_basics.rst:
```diff
@@ -45,61 +48,104 @@ C variable and type definitions
 
-    Structs can be declared as ``cdef packed struct``, which has
-    the same effect as the C directive ``#pragma pack(1)``.
+        .. NOTE:: Currently, Pure Python mode does not support enums.
```
At least, I haven't found any reference to it :-).

0dminnimda replied:
> At least, I haven't found any reference to it :-).

It was worth trying to write enum, and what if it's .. `cython.enum` .., but by chance it's just not documented, lol 🙃

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -45,61 +48,104 @@ C variable and type definitions
+                cdef struct Grail *gp # WRONG
 
-        ctypedef int* IntPtr
+            There is also a ``ctypedef`` statement for giving names to types, e.g.::
```
`cython.typedef`

matusvalo replied:
done, I missed that. Thanks!

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -45,61 +48,102 @@ C variable and type definitions
+
+        .. code-block:: python
+
+            @cython.cfunc
+            def eggs(l: cython.ulong, f: cython.float) -> cython.int:
+                ...
+
+    .. group-tab:: Cython
+
+        .. code-block:: cython
+
+            cdef int eggs(unsigned long l, float f):
+                ...
```
The versions will not align here, it is easier to achieve by moving the code out to the file and adding a space to the line before the `cdef` line.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -209,28 +285,78 @@ there is a tiny overhead in calling a :keyword:`cpdef` method from Cython compar
+
+.. tabs::
+
+    .. group-tab:: Pure Python
+
+        .. code-block:: python
+
+         def chips(t: (cython.long, cython.long, cython.double)) -> (cython.int, cython.float)
+            ...
+
+    .. group-tab:: Cython
+
+        .. code-block:: cython
+
+            cdef (int, float) chips((long, long, double) t):
+                ...
```
indentation is missing in pure python

matusvalo replied:
done.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -209,28 +285,78 @@ there is a tiny overhead in calling a :keyword:`cpdef` method from Cython compar
+
+        .. code-block:: python
+
+            def spam(python_i, python_s):
+                i: cython.int = python_i
+                s: cython.p_char = python_s
+                ...
+
+    .. group-tab:: Cython
+
+        .. code-block:: cython
+
+            def spam(python_i, python_s):
+                cdef int i = python_i
+                cdef char* s = python_s
+                ...
```
You should check that they are identical after cythonization, because as far as I can remember, you need to use `cython.locals` for this.

scoder replied:
Why shouldn't annotations work here? They seem to explain the idea well.

0dminnimda replied:
> Why shouldn't annotations work here? They seem to explain the idea well.

I'm not talking about the idea, I said that in this case the annotations may not work as we want, and therefore I said to check if the cythonizaed versions correspond, if so, then there is nothing to worry about.

scoder replied:
Annotations are fine.

0dminnimda replied:
@scoder Cython version additionally generates
```c
  __pyx_codeobj__2 = (PyObject*)__Pyx_PyCode_New(2, 0, 0, 4, 0, CO_OPTIMIZED|CO_NEWLOCALS, __pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple_, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_kp_s_test_pyx, __pyx_n_s_spam, 5, __pyx_empty_bytes); if (unlikely(!__pyx_codeobj__2)) __PYX_ERR(0, 5, __pyx_L1_error)
```
Here easier to see:

```c
  __pyx_codeobj__2 = (PyObject*)__Pyx_PyCode_New(2, 0, 0, 4, 0, CO_OPTIMIZED|CO_NEWLOCALS,
__pyx_empty_bytes, __pyx_empty_tuple, __pyx_empty_tuple, __pyx_tuple_, __pyx_empty_tuple,
__pyx_empty_tuple, __pyx_kp_s_test_pyx, __pyx_n_s_spam, 5, __pyx_empty_bytes);
if (unlikely(!__pyx_codeobj__2)) __PYX_ERR(0, 5, __pyx_L1_error)
```

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -274,10 +414,24 @@ parameters and a new reference is returned).
+                ...
 
 declares a parameter called int which is a Python object. You can also use
 object as the explicit return type of a function, e.g.::
```
there are no tabs below

matusvalo replied:
Here are no tabs because there is an issue in cython not allowing moving this code to pure python: #2529. I suppose we need a not there about it.

matusvalo replied:
Added note about bug.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -315,13 +469,22 @@ the implementation (in ``.pyx`` files).
 When in a ``.pyx`` file, the signature is the same as it is in Python itself:
 
-.. literalinclude:: ../../examples/userguide/language_basics/optional_subclassing.pyx
+.. tabs::
+
+    .. group-tab:: Pure Python
+
+        .. literalinclude:: ../../examples/userguide/language_basics/optional_subclassing.py
+
+    .. group-tab:: Cython
+
+        .. literalinclude:: ../../examples/userguide/language_basics/optional_subclassing.pyx
 
 When in a ``.pxd`` file, the signature is different like this example: ``cdef foo(x=*)``.
 This is because the program calling the function just needs to know what signatures are
 possible in C, but doesn't need to know the value of the default arguments.:
 
 .. literalinclude:: ../../examples/userguide/language_basics/optional_subclassing.pxd
+    :caption: optional_subclassing.pxd
```
py and pyx must also have captions, as they are in pxd

matusvalo replied:
done.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -383,10 +546,24 @@ propagating the exception to its caller.
-        ...
+.. tabs::
+
+    .. group-tab:: Pure Python
+
+        .. code-block:: python
+
+            @cython.exceptval(-1)
+            def spam() -> cython.int:
+                ...
+
+    .. group-tab:: Cython
+
+        .. code-block:: cython
+
+            cdef int spam() except -1:
+                ...
```
same as [this](https://github.com/cython/cython/pull/4242/files#r660140212)

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +581,48 @@ returns small results.
+
+    .. group-tab:: Pure Python
+
+        .. code-block:: python
+
+            @cython.exceptval(-1, check=True)
+            def spam() -> cython.int:
+                ...
+
+    .. group-tab:: Cython
+
+        .. code-block:: cython
+
+            cdef int spam() except? -1:
+                ...
```
same as [this](https://github.com/cython/cython/pull/4242/files#r660140212)

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +581,48 @@ returns small results.
+            cdef int spam() except? -1:
+                ...
 
 The "?" indicates that the value ``-1`` only signals a possible error. In this
```
there should be text about the `check` attribute

scoder replied:
… and the text would best be placed in the tab to make it change with the syntax display.

matusvalo replied:
done.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +581,48 @@ returns small results.
+
+    .. group-tab:: Pure Python
+
+        .. code-block:: python
+
+            @cython.exceptval(check=True)
+            def spam() -> cython.int:
+                ...
+
+    .. group-tab:: Cython
+
+        .. code-block:: cython
+
+            cdef int spam() except *:
+                ...
```
same as [this](https://github.com/cython/cython/pull/4242/files#r660140212)

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,11 +837,25 @@ be careful what you do.
 ------------
 
-Where C uses ``"("`` and ``")"``, Cython uses ``"<"`` and ``">"``. For example::
+Where C uses ``"("`` and ``")"``, Cython uses ``"<"`` and ``">"``. For example
```
Pure python text should talk about `cython.cast`

matusvalo replied:
done

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,11 +837,25 @@ be careful what you do.
+            p = <char*>q
 
 When casting a C value to a Python object type or vice versa,
 Cython will attempt a coercion. Simple examples are casts like ``<int>pyobj``,
```
[same](https://github.com/cython/cython/pull/4242/files#r660170813)

0dminnimda commented on docs/examples/userguide/language_basics/casting_python.pxd:
```diff
@@ -0,0 +1,2 @@
+cdef extern from *:
+    ctypedef Py_ssize_t Py_intptr_t
```
The ugly way to make this work in pure python is to use [numpy initialization](https://github.com/cython/cython/blob/efdf996dced075682d86b7e53a12b56b5ebfea38/Cython/Includes/numpy/__init__.pxd#L28).
i.e. `from numpy cimport Py_intptr_t`

But, since this tutorial does not focus on the `extern` syntax, it is not clear what it is and how it works, so we can move this type definition somewhere in `Cython/Includes/cpython`

What do you think @scoder?

scoder replied:
Add it to the includes, definitely.
We could also add it to the known C types.

0dminnimda replied:
In this case, the pyx version will also need to use the new `cimport`

0dminnimda replied:
@scoder I would generally suggest covering all of `Python.h` with type declarations. Not necessarily in this PR, but it will be more pleasant if there are already `ctypedef `s for the cpython.

scoder replied:
> @scoder I would generally suggest covering C API path of `Python.h` with type declarations. Not necessarily in this PR, but it will be more pleasant if there are already `ctypedef `s for the cpython.

I fail to understand what you are suggesting here.

0dminnimda replied:
> I fail to understand what you are suggesting here.

I suggested to create [these](https://github.com/cython/cython/tree/master/Cython/Includes/cpython) includes, but it seems that many of them are already listed in those pxd-s.

0dminnimda commented on docs/examples/userguide/language_basics/casting_python.py:
```diff
@@ -0,0 +1,22 @@
+
+    python_string = "foo"
+
+    # Note that the variables below are automatically inferred
+    # as the correct pointer type that is assigned to them.
+    # They do not need to be typed explicitly.
```
Files are not aligned due to comments

0dminnimda replied:
I would suggest that you move these comments outside the code.

scoder replied:
I also wonder if we should just rely on type inference in both versions. And then really move the comment into the preceding text section.

matusvalo replied:
> I also wonder if we should just rely on type inference in both versions.

Is it possible to use type inference in .pyx version? I am getting error when removing `cdef + TYPE` part:
```cython
from cpython.ref cimport PyObject

cdef extern from *:
    ctypedef Py_ssize_t Py_intptr_t

python_string = "foo"

ptr = <void*>python_string
adress_in_c = <Py_intptr_t>ptr
address_from_void = adress_in_c        # address_from_void is a python int

ptr2 = <PyObject*>python_string
address_in_c2 = <Py_intptr_t>ptr2
address_from_PyObject = address_in_c2  # address_from_PyObject is a python int

assert address_from_void == address_from_PyObject == id(python_string)

print(<object>ptr)                     # Prints "foo"
print(<object>ptr2)                    # prints "foo"
```
This code is failing to compile with following error:

```
ptr2 = <PyObject*>python_string
      ^
------------------------------------------------------------

test.pyx:12:7: Cannot convert 'PyObject *' to Python object
```

Maybe I understood it incorrectly.

scoder replied:
Type inference only applies to functions, not the module global code (which only has Python objects in a dict, except for explicitly declared C variables).

matusvalo replied:
Hmm so I don't understand. I am still learning cython so I don't understand everything, but it means that I should move cython language version to `main()` function in order to make variables local?

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -657,10 +939,25 @@ direct equivalent in Python.
 * There is an ``&`` operator, with the same semantics as in C.
 * The null C pointer is called ``NULL``, not ``0`` (and ``NULL`` is a reserved word).
-* Type casts are written ``<type>value`` , for example,::
+* Type casts are written ``<type>value`` , for example,
```
[same](https://github.com/cython/cython/pull/4242/files#r660170813)

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -657,10 +939,25 @@ direct equivalent in Python.
 * There is an ``&`` operator, with the same semantics as in C.
 * The null C pointer is called ``NULL``, not ``0`` (and ``NULL`` is a reserved word).
-* Type casts are written ``<type>value`` , for example,::
+* Type casts are written ``<type>value`` , for example,
```
[same](https://github.com/cython/cython/pull/4242/files#r660170813)

matusvalo replied:
done

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -657,10 +939,25 @@ direct equivalent in Python.
+
+          .. code-block:: python
+
+              p: cython.p_char
+              q: cython.p_float
+              p = cython.cast(cython.p_char, q)
+
+      .. group-tab:: Cython
+
+          .. code-block:: cython
+
+              cdef char* p, float* q
+
+              p = <char*>q
```
take a look [here](https://github.com/cython/cython/pull/4242/files#r660136709)

0dminnimda replied:
Alternatively, you can split `cdef` across two lines.

matusvalo replied:
cdef splitted across two lines.

0dminnimda commented on docs/examples/quickstart/cythonize/integrate_cy.pyx:
```diff
@@ -2,6 +2,8 @@ def f(double x):
     return x ** 2 - x
 
 
+
+
```
?

scoder replied:
Probably makes space for decorators in the Python version?

matusvalo replied:
Yes I have tried to align the two source code as requested in issue.

0dminnimda replied:
hmm it's 1) out of the scope of this tutorial, 2) second file already was aligned, look [here](https://github.com/matusvalo/cython/blob/b1067464260957c69d6756e77a726c71183d2569/docs/examples/quickstart/cythonize/integrate.py) and [here](https://github.com/matusvalo/cython/blob/b1067464260957c69d6756e77a726c71183d2569/docs/examples/quickstart/cythonize/integrate_cy.pyx)

0dminnimda replied:
so i still don't understand

matusvalo replied:
I doublechecked that and it was introduced by mistake. Fixed.

matusvalo commented:
Thank you @scoder, @0dminnimda for your review. I will try to fix your comments.

matusvalo commented:
I have also question - is it possible to rewrite pointer to function in pure python?
```
int (*grail)(int, char*) except -1
```

To be honest, I am not sure how ... If not I will add a note about it.

0dminnimda commented:
It will also be very good if you attach a links to correspondent issues to the notes (in fact I think warnings are better in such cases) about the unrealized functionality, this will speed up the solution process.

0dminnimda commented on docs/examples/userguide/language_basics/parameter_refcount.py:
```diff
@@ -0,0 +1,23 @@
+    refcount = obj.ob_refcnt
+    print('Inside borrowed_reference: {refcount}'.format(refcount=refcount))
+
+def main():
+    print('Initial refcount: {refcount}'.format(refcount=python_dict_refcount))
+    owned_reference(python_dict)
+    borrowed_reference(cython.cast(cython.pointer(PyObject), python_dict))
```
the original script does not have this function, it must be module level code

0dminnimda replied:
In main we don't declare anything, we can throw out main in this case

scoder commented:
> I have also question - is it possible to rewrite pointer to function in pure python?
> ```
> int (*grail)(int, char*) except -1
> ```
> To be honest, I am not sure how ... If not I will add a note about it.

I don't think there currently is a way. I personally never needed this, but it seems to be lacking, so I created a ticket for it. https://github.com/cython/cython/issues/4279

matusvalo commented:
@scoder @0dminnimda I would like to kindly ask for another round of review. I have

* I have adjusted the text to support also pure python code.
* implemented several comments except aligning the lines between cython code and pure python mode ( I would like to do it in the end)

In general, it was pretty difficult to write documentation supporting both types. I am afraid that the text will become more difficult to read. In general I am missing in sphinx-tabs feature putting special version of text based on tab selected. This will allow specialize more text to cython/pure python version. 

So please could you go over the text and give me a feedback whether it is clear and readable or give me comments how it can be improved?

Another note: I have introduced two terms: C function/C method and hybrid function/hybrid method to get rid of `cdef` and `cpdef` mentions across the text. I suppose it would be great to introduce Glossary to Cython documentation as it is in python [1].

Moreover, there is more open points in the documentation, that I need also a help with - e.g.:

https://github.com/cython/cython/blame/master/docs/src/userguide/language_basics.rst#L464


[1] https://docs.python.org/3/glossary.html

0dminnimda commented:
> I am missing in sphinx-tabs feature putting special version of text based on tab selected. This will allow specialize more text to cython/pure python version.

If I understand you correctly, you should know that there are no restrictions that force you to insert only Include into tabs, so you should put specialized text in tabs, look [here](https://github.com/cython/cython/pull/4226/commits/dba1f36578a1acdeef965857570a989ac4a40610) for example

You can see how it works [here](https://cython.readthedocs.io/en/latest/src/tutorial/cython_tutorial.html#primes):

![IMG_20210713_233103](https://user-images.githubusercontent.com/52697657/125521639-14a54598-7939-4160-b18e-d4a9392bc8c3.jpg)
![IMG_20210713_233040](https://user-images.githubusercontent.com/52697657/125521643-ff51db27-9a93-4ea2-833a-181789af75f6.jpg)

0dminnimda commented:
> C function/C method

It doesn't have to be C, it can be C ++, maybe something else. It is better to remove the indication of the language from the term, it will confuse readers.

0dminnimda commented:
> I suppose it would be great to introduce Glossary to Cython documentation

There is a [glossary](https://cython.readthedocs.io/en/latest/src/userguide/glossary.html), although there are not very many terms there.

da-woods commented:
> > C function/C method
> 
> It doesn't have to be C, it can be C ++, maybe something else. It is better to remove the indication of the language from the term, it will confuse readers.

I'm not terribly keen on the term but for different reasons:

1. it implies that *only* `cdef` or `cpdef` functions are compiled into C, where as actually regular `def` functions/methods are compiled too and the internals look close to identical. It's really only the calling convention that changes. Lots of people get hung up on making everything `cdef` because they think it does more than it actually does.
2. it's easily confused with `extern` C functions.

With that said, I don't have a better suggestion right now.

0dminnimda commented:
> e.g.:
> 
> https://github.com/cython/cython/blame/master/docs/src/userguide/language_basics.rst#L464

If you wondering what `extern` is doing here, look [here](https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html?highlight=extern#external-declarations) 

P.S a tricky way to point to a line in rst ;)

0dminnimda commented:
By the way, they use a function pointer, so there is nothing to worry about until this functionality is implemented.

0dminnimda commented on docs/examples/userguide/extension_types/shrubbery.pyx:
```diff
@@ -1,8 +1,10 @@
 
+
 cdef class Shrubbery:
     cdef int width, height
```
In pure python version the values ​​are on different lines, it is better to align them by compressing there or splitting here

matusvalo replied:
done thank you!

0dminnimda commented:
> 1. it implies that _only_ `cdef` or `cpdef` functions are compiled into C, where as actually regular `def` functions/methods are compiled too and the internals look close to identical. It's really only the calling convention that changes. Lots of people get hung up on making everything `cdef` because they think it does more than it actually does.

Good apporunity to point this in the documentation!

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,137 @@ the use of ‘early binding’ programming techniques.
+
+    .. group-tab:: Pure Python
+
+        .. literalinclude:: ../../examples/userguide/language_basics/struct_union_enum.py
```
I don't think we should show anything except the note (better to be warning), moreover structs, thought there are close but not substitutes for each other

0dminnimda commented:
Ok, I see the need of universal term that will represent things defined/declared by `cdef`/`declare`/cython annotations, because it's weird to call `a: cython.float` as "cdef variable".

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,137 @@ the use of ‘early binding’ programming techniques.
+
+        .. code-block:: cython
+
+            cdef int i, j, k
```
Still misaligned with pure python version

matusvalo replied:
The code is not 100% aligned everywhere. I have focused mainly finish the text around the code tu support both versions.

scoder replied:
I like how this shows off one of the main advantages of the Cython syntax: it's very concise.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,137 @@ the use of ‘early binding’ programming techniques.
 
-See also :ref:`struct-union-enum-styles`
+            global_x = declare(cython.int)
+            def main():
```
Why global_x only added to pure version and content of main function must be module-level

matusvalo replied:
`global_x` was added just to show how to declare global variables (using `declare()` function).

0dminnimda replied:
You could use declare in both versions or use cdef in cython version instead of declare, i think I'll look better

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,137 @@ the use of ‘early binding’ programming techniques.
+
+        .. literalinclude:: ../../examples/userguide/language_basics/struct_union_enum.py
+
+        .. NOTE:: Currently, Pure Python mode does not support enums.
```
Would be good if all notes will be formatted equally.

scoder replied:
```suggestion
        .. note:: Currently, Pure Python mode does not support enums.
```

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +187,88 @@ You can read more about them in :ref:`extension-types`.
+        at compile time for stack allocated arrays. Cython doesn't support variable length arrays from C99.
+        Note that Cython uses array access for pointer dereferencing.
+
+    .. group-tab:: Cython
```
Large portion of this tab is copy pasting, so i suggest to move same text out and use tabs only for different sections

matusvalo replied:
Tabs were introduced by mistake here. I have removed them.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +187,88 @@ You can read more about them in :ref:`extension-types`.
+        Arrays use the normal C array syntax, e.g. ``int[10]``, and the size must be known
+        at compile time for stack allocated arrays. Cython doesn't support variable length arrays from C99.
+        Note that Cython uses array access for pointer dereferencing, as ``*x`` is not valid Python syntax,
+        whereas ``x[0]`` is.
```
I don't get this addiction, can you explain what do you mean by this, please?
", as ``*x`` is not ..."

scoder replied:
The sentence seems correct and clear to me.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +187,88 @@ You can read more about them in :ref:`extension-types`.
+        containers.
+
+        Pointer types are constructed by appending a ``p_`` to the base type
+        they point to, e.g. ``pp_int`` for a pointer to a pointer to a C int.
```
`by needed number of "p"s and then "_"` Would be great if you shaped this sentence like that, feel free to change it for better fit.

matusvalo replied:
Done thank you!

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +187,88 @@ You can read more about them in :ref:`extension-types`.
+
+        Pointer types are constructed by appending a ``p_`` to the base type
+        they point to, e.g. ``pp_int`` for a pointer to a pointer to a C int.
+        Further pointer types can be constructed with ``pointer()`` function.
```
It's false statement, there are "ppp_"  in shadow.py + very important mention is that it's only for interpretation mode, in the compilation mode there are no limit

scoder replied:
That's not important at all. In real code, pointers will almost never have more than two indirection levels, let alone more than three. Let's be reasonable.

0dminnimda replied:
@scoder, yeah, you probably would never want to use more then two pointers, but documentation should not lie, at least pp should be changed to ppp, so documentation would convey truth

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +187,88 @@ You can read more about them in :ref:`extension-types`.
+        containers.
+
+        Pointer types are constructed by appending a ``p_`` to the base type
+        they point to, e.g. ``pp_int`` for a pointer to a pointer to a C int.
```
`cython.pp_int`

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,137 @@ the use of ‘early binding’ programming techniques.
+                 ULong = typedef(cython.ulong)
 
-        ctypedef int* IntPtr
+                 IntPtr = typedef(cython.p_int)
```
`cython.typedef`

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,137 @@ the use of ‘early binding’ programming techniques.
-    pointing to a ``Grail`` you would write::
 
-        cdef Grail *gp
+    There is also support for giving names to types by ``ctypedef`` statement or ``typedef()`` function, e.g.
```
We could move this line to the tabs, but idk is this that necessary

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,137 @@ the use of ‘early binding’ programming techniques.
-The :keyword:`cdef` statement is used to declare C variables, either local or
-module-level::
+C variables can be declared by using the :keyword:`cdef` statement, by annotating
+variable by special cython type or using function ``declare()``. Statement :keyword:`cdef`
```
It will be better to put them into list like
- one
- two
- three

It's not easy to understand this first time reading

matusvalo replied:
done.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +187,88 @@ You can read more about them in :ref:`extension-types`.
+        .. code-block:: python
+
+            def main():
+                foo: list = []
```
Must he module-level, please remove main

matusvalo replied:
@0dminnimda to be honest I don't understand. @scoder mentioned that annotated variables should be moved to function since Cython does not support annotated global variables:

https://github.com/cython/cython/pull/4242#discussion_r660033760
https://github.com/cython/cython/pull/4242#pullrequestreview-694259168

Hence, I put all code to `main()` function. Or better idea is to use `declare()`? I am not sure what is better :-(

scoder replied:
Not sure if `main()` is a good name, but it's definitely good enough for now and easy to change later. Let's stick with it.

0dminnimda replied:
> Or better idea is to use declare()?

that's the other way of doing this

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +187,88 @@ You can read more about them in :ref:`extension-types`.
-respectively, as statically typing variables with these Python
-types has zero advantages.
+
+.. Note:: The Python types ``int``, ``long``, and ``float`` are not available for static
```
Why you chose to add note here?

scoder replied:
Seems a good change to me.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +187,88 @@ You can read more about them in :ref:`extension-types`.
 
-    cdef (double, int) bar
+            def main():
+                bar: (cython.double, cython.int)
```
Must be module-level, please remove main

matusvalo replied:
See https://github.com/cython/cython/pull/4242#discussion_r669362405

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -178,8 +294,11 @@ Grouping multiple C declarations
 If you have a series of declarations that all begin with :keyword:`cdef`, you
 can group them into a :keyword:`cdef` block like this:
 
+.. note:: This is supported only by Cython language.
```
Other option is to use multiple `declare` call it's the option and looks pretty similar

matusvalo replied:
`declare()` can be used for multiple variable declaration but the scope of cdef block is more generic (functions, structs etc).

0dminnimda replied:
multiline* sorry

0dminnimda replied:
> but the scope of cdef block is more generic (functions, structs etc).

Yeah true, didn't think about it

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +308,98 @@ Python functions vs. C functions
 :term:`Python objects<Python object>` as parameters and return Python objects.
 
-C functions are defined using the new :keyword:`cdef` statement. They take
+C functions are defined using the new :keyword:`cdef` statement in Cython language or by ``@cfunc`` decorator. They take
```
Better use tabs for this purpose instead

0dminnimda replied:
Apples to other similar cases

matusvalo replied:
Closing based on following comment: https://github.com/cython/cython/pull/4242#issuecomment-879666879

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +308,98 @@ Python functions vs. C functions
+        .. code-block:: python
+
+            @cfunc
+            def chips(t: (cython.long, cython.long, cython.double)) -> (cython.int, cython.float):
```
Probably again will not align

0dminnimda commented:
Will left [this](https://github.com/cython/cython/pull/4242/files#r669204086) here so @scoder can notice

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -257,10 +426,24 @@ Python objects as parameters and return values
+
+        .. code-block:: cython
+
+            cdef spamobjs(x, y):
```
Will not align

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -274,25 +457,50 @@ parameters and a new reference is returned).
-declares a parameter called int which is a Python object. You can also use
+        .. code-block:: cython
+
+            cdef ftang(object int):
```
Will not align

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -374,19 +593,33 @@ through defined error return values.  For functions that return a Python object
+
+        .. code-block:: cython
+
+            cdef int spam() except -1:
```
Will not align

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +637,56 @@ returns small results.
+
+        The ``check=True`` indicates that the value ``-1`` only signals a possible error.
+        In this
+        case, Cython generates a call to :c:func:`PyErr_Occurred` if the exception value
```
no need in new line here and in corresponding tab

matusvalo replied:
fixed.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +637,56 @@ returns small results.
+        The ``?`` indicates that the value ``-1`` only signals a possible error.
+        In this
+        case, Cython generates a call to :c:func:`PyErr_Occurred` if the exception value
+        is returned, to make sure it really received an exception and not just a normal
```
Please, move the same text out of tabs

matusvalo replied:
fixed.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +637,56 @@ returns small results.
+
+        .. code-block:: cython
+
+            cdef int spam() except *:
```
Will not align

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +637,56 @@ returns small results.
+
+        .. code-block:: python
+
+            @exceptval(check=True)
```
You forgot `cython.cfunc` and same in other places

0dminnimda replied:
But I doubt it a little, so you better double check

matusvalo replied:
Done thank you!

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -430,14 +699,16 @@ An external C++ function that may raise an exception can be declared with::
 
     cdef int spam() except +
 
+.. note:: Pure python mode currently does not have equivalent to `except +`.
```
Created issue for that #4282

0dminnimda replied:
```suggestion
.. note:: Pure python mode currently does not have equivalent to `except +`. (Github issue :issue:`4282`)
```

0dminnimda replied:
And also it would be really great if you will add those references to all other notes that needs that.

0dminnimda replied:
As said in #4282 we should not allow it, and it should be reflected in the text + this conversation should be resolved

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -548,10 +844,24 @@ as the C string is needed. If you can't guarantee that the Python string will
+
+            def main():
+                s: cython.p_char
+                s = pystring1 + pystring2
```
Again, this should be module-level and there should be no main

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -562,11 +872,26 @@ leaving ``s`` dangling. Since this code could not possibly work, Cython refuses
+            def main():
+                s: cython.p_char
+                p = pystring1 + pystring2
+                s = p
```
this should be module-level and there should be no main

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +906,93 @@ be careful what you do.
 
-Where C uses ``"("`` and ``")"``, Cython uses ``"<"`` and ``">"``. For example::
+Cython language supports type casting in a simmilar way as C. Where C uses ``"("`` and ``")"``,
+Cython uses ``"<"`` and ``">"``. In pure python mode, ``cast()`` function is used. For example:
```
`cython.cast`
You may consider moving this to tabs too

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +906,93 @@ be careful what you do.
+        with ``cast(object, ...)``, or a more specific builtin or extension type
+        (e.g. ``cast(MyExtType, ptr)``). This will increase the reference count of
+        the object by one, i.e. the cast returns an owned reference.
+        Here is an example:
```
The only place as far where you should not move test, yeah it'll be better it you could turn off tabs but will left finctionality of changing content. (I think it's good issue for sphinx-tabs repo)

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +906,93 @@ be careful what you do.
 ``MyExtensionType`` without any checking at all.
 
-To have a cast checked, use the syntax like: ``<MyExtensionType?>x``.
+To have a cast checked, use the syntax like: ``<MyExtensionType?>x`` in Cython language
```
Will be good to move to tabs

0dminnimda replied:
Just as reminder, i wil not comment on every one like those

matusvalo replied:
Based on https://github.com/cython/cython/pull/4242#issuecomment-879666879 I would close it without change.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -651,16 +1025,35 @@ direct equivalent in Python.
-* The null C pointer is called ``NULL``, not ``0`` (and ``NULL`` is a reserved word).
-* Type casts are written ``<type>value`` , for example,::
+* There is an ``&`` operator in Cython language, with the same semantics as in C. In pure python mode use ``address()`` function.
+* The null C pointer is called ``NULL``, not ``0``. ``NULL`` is a reserved word in Cython language
```
`cython.NULL` for pure python

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -651,16 +1025,35 @@ direct equivalent in Python.
+                  p: cython.p_char
+                  q: cython.p_float
+
+                  p = cast(cython.p_char, q)
```
You know, main and stuff

0dminnimda commented:
UFF, it was an investment of time that I hope will pay off
Well, mostly well done!

0dminnimda commented:
I didn't comment on ones with changed terminology, because this is still in process of approval, but when everything will be defined those names also should be changed

0dminnimda commented:
> UFF, it was an investment of time that I hope will pay off

Around 1000 lines covered at 7 hours of review.
Hmm what could it say .. am i reviewing slow? Or maybe I'm just I am a diligent person?
Well, it looks like the answers to these questions will not be found for millennia ... We all need to prepare, waiting will not be easy :)

mattip commented:
> Around 1000 lines covered at 7 hours of review.

Perhaps it would be better to iterate on these PRs. Get a first version merged that is "good enough", and then do another PR to clean up formatting things like alignment, bullets, notes, etc. The pattern of over 100 comments on a documentation PR is unsustainable and requires too much work to resolve the comments.

matusvalo commented:
> Around 1000 lines covered at 7 hours of review.

Thank you very much for your hard work! This PR is rather difficult and needs multiple iterations to get it right.

> If I understand you correctly, you should know that there are no restrictions that force you to insert only include into tabs, so you should put specialized text in tabs, look here for example

Regarding my note about sphinx-tabs. I meant that it would be great to have two versions of text **without** putting tabs before them. It means that by switching tabs also some portions of text would change that are not placed inside a tab.

> It doesn't have to be C, it can be C ++, maybe something else. It is better to remove the indication of the language from the term, it will confuse readers.

I agree that the naming convention is not ideal. I came up with C-function / hybrid function only since it was already used in text before. But in order to make text not tightly connected to Cython language we need to move from `cdef`, `cpdef` terms. @scoder could you help us with that?

scoder commented:
>Regarding my note about sphinx-tabs. I meant that it would be great to have two versions of text **without** putting tabs before them. It means that by switching tabs also some portions of text would change that are not placed inside a tab.

I think it's good to give hints from time to time that there *are* syntax variants. We should not make normal text change, and should generally keep the amount of text in tabs low.

If that makes certain text sections confusing, then there's probably a better way to present their meaning.

Matti is right, we should concentrate on the major changes first, and then iterate. Many things will become clearer (and also more visible to others) once there is a nicely formatted version up on docs.cython.org.

scoder commented:
>+* The null C pointer is called ``NULL``, not ``0``. ``NULL`` is a
>reserved word in Cython language
>
>`cython.NULL` for pure python

I think saying NULL is fine. We should just make sure that it's clear from the beginning that the „cython” module is where you can find such C names.

scoder commented:
>-    cdef char *s
>-    p = pystring1 + pystring2
>-    s = p
>+    .. group-tab:: Pure Python
>+
>+        .. code-block:: python
>+
>+            def main():
>+                s: cython.p_char
>+                p = pystring1 + pystring2
>+                s = p
>
>this should be module-level and there should be no main

As long as the function and module scopes behave differently in Cython (and in Python), I'd rather stick to the scope that most users will write real code in, which is the function scope.

scoder commented:
>+.. note:: Pure python mode currently does not have equivalent to
>`except +`.

And it also does not need it, right? So there's no „currently”, it's not a limitation, it's probably going to stay that way. That's worth making clear - for external declarations, use a .pxd file. Always.

scoder commented:
>-C functions are defined using the new :keyword:`cdef` statement. They
>take
>+C functions are defined using the new :keyword:`cdef` statement in
>Cython language or by ***@***.***`` decorator. They take
>
>Better use tabs for this purpose instead

No, I'd rather see both mentioned. It's good to give readers the mental link between both syntax variants, because they will quite likely run into both in the future.

scoder commented:
> Ok, I see the need of universal term that will represent things defined/declared by `cdef`/`declare`/cython annotations, because it's weird to call `a: cython.float` as "cdef variable".

The term "typed variable" would probably be ok in most cases. Maybe not all, but I think that's the main idea when we refer to a `cdef` variable.

0dminnimda commented:
> As long as the function and module scopes behave differently in Cython (and in Python), I'd rather stick to the scope that most users will write real code in, which is the function scope

Sure, then it should be in both cython and pure python versions, so there would not be any strange difference, that user can think is the only way

0dminnimda commented:
> > As long as the function and module scopes behave differently in Cython (and in Python), I'd rather stick to the scope that most users will write real code in, which is the function scope
> 
> Sure, then it should be in both cython and pure python versions, so there would not be any strange difference, that user can think is the only way for this version

0dminnimda commented:
> I think saying NULL is fine. We should just make sure that it's clear from the beginning that the „cython” module is where you can find such C names.

Kinda yeah, but it's not hard to add `cython.` and easier for reader to understand

0dminnimda commented:
> The term "typed variable" would probably be ok in most cases. Maybe not all, but I think that's the main idea when we refer to a `cdef` variable.

Ok then it should replace "C variables/functions/methods"

0dminnimda commented:
> Perhaps it would be better to iterate on these PRs. Get a first version merged that is "good enough", and then do another PR to clean up formatting things like alignment, bullets, notes, etc. The pattern of over 100 comments on a documentation PR is unsustainable and requires too much work to resolve the comments.

I would propose to limit number of new lines per rst (one rst per PR) one could use, for example 200, or limit number of chapters that can be covered, for example 2.
And write about it in the original ticket, so one could be ready to split their PR it they want to cover whole document.

0dminnimda commented:
> The term "typed variable" would probably be ok in most cases. Maybe not all, but I think that's the main idea when we refer to a `cdef` variable

What's about `cpdef`s?

0dminnimda commented:
> Sure, then it should be in both cython and pure python versions, so there would not be any strange difference, that user can think is the only way for this version

Or just add empty line in front of cython version + check that there's some text that explains why we using function in pure python

0dminnimda commented:
Ok, @matusvalo i cannot resolve my reviews, so feel free to do this in cases like main

matusvalo commented:
> i cannot resolve my reviews, so feel free to do this in cases like main

Let's wait for @scoder review when will have time. I tried to fix as many issues as possible.

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -178,8 +283,11 @@ Grouping multiple C declarations
 If you have a series of declarations that all begin with :keyword:`cdef`, you
 can group them into a :keyword:`cdef` block like this:
 
+.. note:: This is supported only by Cython language.
```
```suggestion
.. note:: This is supported only in Cython's ``cdef`` syntax.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +297,98 @@ Python functions vs. C functions
 There are two kinds of function definition in Cython:
 
-Python functions are defined using the def statement, as in Python. They take
+Python functions are defined using the ``def`` statement, as in Python. They take
```
```suggestion
Python functions are defined using the :keyword:`def` statement, as in Python. They take
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +297,98 @@ Python functions vs. C functions
 :term:`Python objects<Python object>` as parameters and return Python objects.
 
-C functions are defined using the new :keyword:`cdef` statement. They take
+C functions are defined using the new :keyword:`cdef` statement in Cython language or by ``@cfunc`` decorator. They take
```
```suggestion
C functions are defined using the :keyword:`cdef` statement in Cython syntax or with the ``@cfunc`` decorator. They take
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +297,98 @@ Python functions vs. C functions
-Cython module must be declared as Python functions using def.
-There is also a hybrid function, called :keyword:`cpdef`. A :keyword:`cpdef`
+Cython module must be declared as Python functions using ``def``.
+There is also a hybrid function, declared by :keyword:`cpdef` in Cython language or by ``@ccall`` decorator. A hybrid function
```
```suggestion
There is also a hybrid function, declared with :keyword:`cpdef` in ``.pyx`` files or with the ``@ccall`` decorator.  These functions
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +297,98 @@ Python functions vs. C functions
-There is also a hybrid function, called :keyword:`cpdef`. A :keyword:`cpdef`
+Cython module must be declared as Python functions using ``def``.
+There is also a hybrid function, declared by :keyword:`cpdef` in Cython language or by ``@ccall`` decorator. A hybrid function
 can be called from anywhere, but uses the faster C calling conventions
```
```suggestion
can be called from anywhere, but use the faster C calling convention
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +297,98 @@ Python functions vs. C functions
+There is also a hybrid function, declared by :keyword:`cpdef` in Cython language or by ``@ccall`` decorator. A hybrid function
 can be called from anywhere, but uses the faster C calling conventions
-when being called from other Cython code. A :keyword:`cpdef` can also be overridden
+when being called from other Cython code. A hybrid function can also be overridden
```
```suggestion
when being called from other Cython code. They can also be overridden
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +297,98 @@ Python functions vs. C functions
 If this happens, most performance gains are of course lost and even if it does not,
-there is a tiny overhead in calling a :keyword:`cpdef` method from Cython compared to
-calling a :keyword:`cdef` method.
+there is a tiny overhead in calling a hybfid method from Cython compared to
```
```suggestion
there is a tiny overhead in calling such a method from Cython compared to
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -243,7 +401,7 @@ with string attributes if they are to be used after the function returns.
 passed in directly using a normal C function call.
 
-Functions declared using :keyword:`cdef` with Python object return type, like Python functions, will return a :keyword:`None`
+C Functions declared using :keyword:`cdef` or ``@cfunc`` decorator with Python object return type, like Python functions, will return a :keyword:`None`
```
```suggestion
C Functions declared using :keyword:`cdef` or the ``@cfunc`` decorator with a
Python object return type, like Python functions, will return a :keyword:`None`
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -274,25 +446,50 @@ parameters and a new reference is returned).
         ...
 
+.. note:: Currently, Cython contains a bug not allowing returning ``object`` in
+    pure python from C function.
```
```suggestion
    pure python from a C function.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -305,23 +502,34 @@ will display::
 
-Unlike C, it is possible to use optional arguments in ``cdef`` and ``cpdef`` functions.
-There are differences though whether you declare them in a ``.pyx``
+Unlike C, it is possible to use optional arguments in C and hybrid functions.
```
```suggestion
Unlike C, it is possible to use optional arguments in C and ``cpdef``/``@ccall`` functions.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -374,19 +582,34 @@ through defined error return values.  For functions that return a Python object
-While this is always the case for :keyword:`def` functions, functions
-defined as :keyword:`cdef` or :keyword:`cpdef` can return arbitrary C types,
+While this is always the case for C functions, functions
+defined as C functions or hybryd functions can return arbitrary C types,
```
```suggestion
defined as C functions or ``cpdef``/``@ccall`` functions can return arbitrary C types,
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -430,14 +686,16 @@ An external C++ function that may raise an exception can be declared with::
 
     cdef int spam() except +
 
+.. note:: Pure python mode does not have equivalent to `except +`.
```
```suggestion
.. note:: These declarations are not used in Python code, only in ``.pxd``  and ``.pyx`` files.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -446,6 +704,8 @@ Some things to note:
 
       int (*grail)(int, char*) except -1
 
+  .. note:: Pointer to function is currently not supported by pure python mode.
```
```suggestion
  .. note:: Pointers to functions are currently not supported by pure Python mode.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -470,23 +730,46 @@ function or a C function that calls Python/C API routines. To get an exception
 
 
-``cpdef`` methods can override ``cdef`` methods:
+Hybrid methods can override C methods:
```
```suggestion
``cpdef``/``@ccall`` methods can override C methods:
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -470,23 +730,46 @@ function or a C function that calls Python/C API routines. To get an exception
 When subclassing an extension type with a Python class,
-``def`` methods can override ``cpdef`` methods but not ``cdef``
-methods:
+C methods can override hybrid methods but not C methods:
```
```suggestion
Python methods can override ``cpdef``/``@ccall`` methods but not plain C methods:
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -562,11 +859,26 @@ leaving ``s`` dangling. Since this code could not possibly work, Cython refuses
 
 The solution is to assign the result of the concatenation to a Python
-variable, and then obtain the ``char*`` from that, i.e.::
+variable, and then obtain the pointer to char from that, i.e.
```
```suggestion
variable, and then obtain the ``char*`` from that, i.e.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +893,93 @@ be careful what you do.
 ------------
 
-Where C uses ``"("`` and ``")"``, Cython uses ``"<"`` and ``">"``. For example::
+Cython language supports type casting in a simmilar way as C. Where C uses ``"("`` and ``")"``,
```
```suggestion
The Cython language supports type casting in a similar way as C. Where C uses ``"("`` and ``")"``,
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +893,93 @@ be careful what you do.
 
-Where C uses ``"("`` and ``")"``, Cython uses ``"<"`` and ``">"``. For example::
+Cython language supports type casting in a simmilar way as C. Where C uses ``"("`` and ``")"``,
+Cython uses ``"<"`` and ``">"``. In pure python mode, ``cast()`` function is used. For example:
```
```suggestion
Cython uses ``"<"`` and ``">"``.  In pure python mode, the ``cython.cast()`` function is used.  For example:
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +893,93 @@ be careful what you do.
+                p = cast(cython.p_char, q)
+
+        When casting a C value to a Python object type or vice versa,
+        Cython will attempt a coercion. Simple examples are casts like ``cast(int, pyobj)``,
+        which converts a Python number to a plain C ``int`` value, or ``cast(bytes, charptr)``,
```
```suggestion
        Cython will attempt a coercion. Simple examples are casts like ``cast(int, pyobj_value)``,
        which converts a Python number to a plain C ``int`` value, or ``cast(bytes, charptr_value)``,
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +893,93 @@ be careful what you do.
 
-The precedence of ``<...>`` is such that ``<type>a.b.c`` is interpreted as ``<type>(a.b.c)``.
+        When casting a C value to a Python object type or vice versa,
+        Cython will attempt a coercion. Simple examples are casts like ``<int>pyobj``,
+        which converts a Python number to a plain C ``int`` value, or ``<bytes>charptr``,
```
```suggestion
        Cython will attempt a coercion. Simple examples are casts like ``<int>pyobj_value``,
        which converts a Python number to a plain C ``int`` value, or ``<bytes>charptr_value``,
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +893,93 @@ be careful what you do.
+        .. literalinclude:: ../../examples/userguide/language_basics/casting_python.py
+            :caption: casting_python.py
+
+        Casting to ``cast(object, ...)`` creates an owned reference. Cython will automatically
```
```suggestion
        Casting with ``cast(object, ...)`` creates an owned reference. Cython will automatically
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +893,93 @@ be careful what you do.
 ``MyExtensionType`` without any checking at all.
 
-To have a cast checked, use the syntax like: ``<MyExtensionType?>x``.
+To have a cast checked, use the syntax like: ``<MyExtensionType?>x`` in Cython language
```
```suggestion
To have a cast checked, use ``<MyExtensionType?>x`` in Cython syntax
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -651,16 +1012,35 @@ direct equivalent in Python.
 * An integer literal is treated as a C constant, and will
   be truncated to whatever size your C compiler thinks appropriate.
   To get a Python integer (of arbitrary precision) cast immediately to
```
```suggestion
  To get a Python integer (of arbitrary precision), cast immediately to
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -651,16 +1012,35 @@ direct equivalent in Python.
-  an object (e.g. ``<object>100000000000000000000``). The ``L``, ``LL``,
-  and ``U`` suffixes have the same meaning as in C.
+  an object (e.g. ``<object>100000000000000000000`` or ``cast(object, 100000000000000000000)``). The ``L``, ``LL``,
+  and ``U`` suffixes have the same meaning in Cython language as in C.
```
```suggestion
  and ``U`` suffixes have the same meaning in Cython syntax as in C.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -651,16 +1012,35 @@ direct equivalent in Python.
-* There is an ``&`` operator, with the same semantics as in C.
-* The null C pointer is called ``NULL``, not ``0`` (and ``NULL`` is a reserved word).
-* Type casts are written ``<type>value`` , for example,::
+* There is an ``&`` operator in Cython language, with the same semantics as in C. In pure python mode use ``address()`` function.
```
```suggestion
* There is an ``&`` operator in Cython, with the same semantics as in C.
  In pure python mode, use the ``cython.address()`` function instead.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -651,16 +1012,35 @@ direct equivalent in Python.
-* The null C pointer is called ``NULL``, not ``0`` (and ``NULL`` is a reserved word).
-* Type casts are written ``<type>value`` , for example,::
+* There is an ``&`` operator in Cython language, with the same semantics as in C. In pure python mode use ``address()`` function.
+* The null C pointer is called ``NULL``, not ``0``. ``NULL`` is a reserved word in Cython language
```
```suggestion
* The null C pointer is called ``NULL``, not ``0``. ``NULL`` is a reserved word in Cython
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -742,6 +1122,8 @@ Python and C, and that Cython uses the Python precedences, not the C ones.
 Integer for-loops
 ------------------
 
+.. note:: This syntax is supported only by Cython language.
```
```suggestion
.. note:: This syntax is supported only in Cython files and not in Python.
```

scoder reviewed (commented):
Lots of grammar issues.

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,143 @@ the use of ‘early binding’ programming techniques.
 
-    cdef int i, j, k
-    cdef float f, g[42], *h
+* using the Cython language :keyword:`cdef` statement,
+* annotating variable by special cython type or
+* using function ``declare()``.
```
```suggestion
* using the Cython specific :keyword:`cdef` statement,
* using PEP-484/526 type annotations with C data types or
* using the function ``cython.declare()``.
```

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -86,7 +86,7 @@ Moreover, C :keyword:`struct`, :keyword:`union` or :keyword:`enum` are supported
 
-        .. NOTE:: Currently, Pure Python mode does not support enums.
+        .. note:: Currently, Pure Python mode does not support enums.
 
```
It's bot the only one, there are other cases that i saw, it's better to replace all note entries in this document with `note`

0dminnimda replied:
i just will find and suggest change, i think it'll be more useful

0dminnimda replied:
This conversation should be resolved

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,143 @@ the use of ‘early binding’ programming techniques.
+* using function ``declare()``.
 
-and C :keyword:`struct`, :keyword:`union` or :keyword:`enum` types:
+Statement :keyword:`cdef` and function ``declare()`` can declare either local or
+module-level variables, but annotated version currently supports only local variables:
```
```suggestion
The :keyword:`cdef` statement and ``declare()`` can define function-local and
module-level variables, but type annotations only affect local variables.
This is because type annotations are not Cython specific, so Cython keeps
the variables in the module dict (as Python values) instead of making them
module internal C variables. Use ``declare()`` in Python code to explicitly
define global C variables.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +193,73 @@ You can read more about them in :ref:`extension-types`.
-Pointer types are constructed as in C, by appending a ``*`` to the base type
-they point to, e.g. ``int**`` for a pointer to a pointer to a C int.
+Pointer types are constructed as in C when using Cython syntax, by appending a ``*`` to the base type
+they point to, e.g. ``int**`` for a pointer to a pointer to a C int. In Pure python mode, pointer types
```
```suggestion
they point to, e.g. ``int**`` for a pointer to a pointer to a C int. In Pure python mode, simple pointer types
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +193,73 @@ You can read more about them in :ref:`extension-types`.
+Pointer types are constructed as in C when using Cython syntax, by appending a ``*`` to the base type
+they point to, e.g. ``int**`` for a pointer to a pointer to a C int. In Pure python mode, pointer types
+use a naming scheme with "p"s instead, separated from the type name with an underscore, e.g. ``cython.pp_int`` for a pointer to
+a pointer to a C int. Further pointer types can be constructed with the ``cython.pointer()`` function.
```
```suggestion
a pointer to a C int.  Further pointer types can be constructed with the ``cython.pointer()`` function,
e.g. ``cython.pointer(cython.int)``.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +193,73 @@ You can read more about them in :ref:`extension-types`.
+    typing in ``.pyx`` files and instead interpreted as C ``int``, ``long``, and ``float``
+    respectively, as statically typing variables with these Python
+    types has zero advantages. On the other hand, annotating in Pure Python with
+    ``int``, ``long``, and ``float`` python types will be interpreted as ``object`` type.
```
```suggestion
    ``int``, ``long``, and ``float`` Python types will be interpreted as
    Python object types.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +299,98 @@ Python functions vs. C functions
+
+            cdef int eggs(unsigned long l, float f):
+                ...
+
+
+
```
```suggestion

```

scoder commented:
Could someone else please take another look at the grammar? I'm sure I didn't find all issues during my review.

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,147 @@ the use of ‘early binding’ programming techniques.
+* using the function ``cython.declare()``.
 
-and C :keyword:`struct`, :keyword:`union` or :keyword:`enum` types:
+The :keyword:`cdef` statement and ``declare()`` can define function-local and
+module-level variables, but type annotations only affect local variables.
```
```suggestion
The :keyword:`cdef` statement and ``declare()`` can define function-local and
module-level variables as well as attributes in classes, but type annotations only
affect local variables and attributes and are ignored at the module level.
```

h-vetinari commented on docs/src/userguide/language_basics.rst:
```diff
@@ -274,25 +453,50 @@ parameters and a new reference is returned).
     cdef object ftang(object int):
         ...
 
+.. note:: Currently, Cython contains a bug not allowing returning ``object`` in
+    pure python from a C function.
```
Is there an issue for this that could be referenced here? That would make it easy to find out if the docs become outdated (by seeing that the referenced issue has been closed).

matusvalo replied:
The bug referenced is following: #2529. Not sure whether we want there references to bugs... (I understand that it adds visibility)

h-vetinari commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +634,53 @@ returns small results.
+            def spam() -> cython.int:
+                ...
+
+        The ``check=True`` indicates that the value ``-1`` only signals a possible error.
```
```suggestion
        The keyword argument ``check=True`` signals that the value ``-1`` _may_ be an error.
```

I can see that the formulation existed before, but I find it hard to parse.

h-vetinari commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +634,53 @@ returns small results.
+            cdef int spam() except? -1:
+                ...
+
+        The ``?`` indicates that the value ``-1`` only signals a possible error.
```
```suggestion
        The ``?`` indicates that the value ``-1`` _may_ be an error.
```

Consistency with the above.

h-vetinari commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +900,93 @@ be careful what you do.
+
+        When casting a C value to a Python object type or vice versa,
+        Cython will attempt a coercion. Simple examples are casts like ``cast(int, pyobj_value)``,
+        which converts a Python number to a plain C ``int`` value, or ``cast(bytes, charptr_value)``,
+        which copies a C ``char*`` string into a new Python bytes object.
```
```suggestion
        which convert a Python number to a plain C ``int`` value, or the statement ``cast(bytes, charptr_value)``,
        which copies a C ``char*`` string into a new Python bytes object.
```

Casts [...] which convert [...] or `<xxx>`** which copies [...]

**  changed object for clarity & readability

h-vetinari commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +900,93 @@ be careful what you do.
-    p = <char*>q
+        To get the address of some Python object, use a cast to a pointer type
+        like ``cast(p_void, ...)`` or ``cast(pointer(PyObject), ...)``.
+        You can also cast a C pointer back to a Python object reference
+        with ``cast(object, ...)``, or a more specific builtin or extension type
+        (e.g. ``cast(MyExtType, ptr)``). This will increase the reference count of
```
```suggestion
        You can also cast a C pointer back to a Python object reference
        with ``cast(object, ...)``, or to a more specific builtin or extension type
        (e.g. ``cast(MyExtType, ptr)``). This will increase the reference count of
```

h-vetinari commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +900,93 @@ be careful what you do.
 
-The precedence of ``<...>`` is such that ``<type>a.b.c`` is interpreted as ``<type>(a.b.c)``.
+        When casting a C value to a Python object type or vice versa,
+        Cython will attempt a coercion. Simple examples are casts like ``<int>pyobj_value``,
+        which converts a Python number to a plain C ``int`` value, or ``<bytes>charptr_value``,
+        which copies a C ``char*`` string into a new Python bytes object.
```
```suggestion
        Cython will attempt a coercion. Simple examples are casts like ``<int>pyobj_value``,
        which convert a Python number to a plain C ``int`` value, or the statement ``<bytes>charptr_value``,
        which copies a C ``char*`` string into a new Python bytes object.
```

As above

h-vetinari commented on docs/src/userguide/language_basics.rst:
```diff
@@ -581,44 +900,93 @@ be careful what you do.
+        To get the address of some Python object, use a cast to a pointer type
+        like ``<void*>`` or ``<PyObject*>``.
+        You can also cast a C pointer back to a Python object reference
+        with ``<object>``, or a more specific builtin or extension type
```
```suggestion
        with ``<object>``, or to a more specific builtin or extension type
```

h-vetinari reviewed (commented):
> @scoder: Could someone else please take another look at the grammar? I'm sure I didn't find all issues during my review.

Did a grammar review of all new text (perhaps not with the finest of combs, but still...), and didn't find much anymore.

scoder commented:
>> +.. note:: Currently, Cython contains a bug not allowing returning
>``object`` in
>+    pure python from a C function.
>
>The bug referenced is following: #2529. Not sure whether we want there
>references to bugs... (I understand that it adds visibility)

I'd rather see it fixed in time and not need a link to it. But that's a wish. We'll see.

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -890,6 +1280,11 @@ Conditional Compilation
+.. note::
+
+    Cython currently does not support conditional compilation and compile-time
+    definitions in Pure Python mode.
```
I propose split this to two notes to each subsection and add references to those issues: #4286, #4287

0dminnimda reviewed (changes requested):
Since you struggled with references, I'll help with that

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -742,6 +1130,8 @@ Python and C, and that Cython uses the Python precedences, not the C ones.
 Integer for-loops
 ------------------
 
+.. note:: This syntax is supported only in Cython files and not in Python.
```
```suggestion
.. note:: This syntax is supported only in Cython files. Use a normal `for-in-range()` loop instead.
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,148 @@ the use of ‘early binding’ programming techniques.
+
+        .. code-block:: cython
+
+            cdef int i, j, k
+            cdef float f, g[42], *h
```
```suggestion
            cdef int a_global_variable

            def func():
                cdef int i, j, k
                cdef float f, g[42], *h
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,148 @@ the use of ‘early binding’ programming techniques.
-.. note::
+        .. code-block:: python
+
+            global_x = declare(cython.int)
+            def main():
```
```suggestion
            a_global_variable = declare(cython.int)

            def func():
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -44,72 +47,148 @@ the use of ‘early binding’ programming techniques.
+
+        .. code-block:: python
+
+            @cfunc
```
```suggestion
            @cython.cfunc
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +306,95 @@ Python functions vs. C functions
+            def spam(i: cython.int, s: cython.p_char):
+                ...
+
+            @cfunc
```
```suggestion
            @cython.cfunc
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -189,48 +306,95 @@ Python functions vs. C functions
+
+        .. code-block:: python
+
+            @cfunc
```
```suggestion
            @cython.cfunc
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -257,10 +422,24 @@ Python objects as parameters and return values
+
+        .. code-block:: python
+
+            @cfunc
```
```suggestion
            @cython.cfunc
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -274,25 +453,50 @@ parameters and a new reference is returned).
+
+        .. code-block:: python
+
+            @cfunc
```
```suggestion
            @cython.cfunc
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -374,19 +589,34 @@ through defined error return values.  For functions that return a Python object
+
+        .. code-block:: python
+
+            @cfunc
+            @exceptval(-1)
```
```suggestion
            @cython.cfunc
            @cython.exceptval(-1)
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +634,53 @@ returns small results.
+
+        .. code-block:: python
+
+            @cfunc
+            @exceptval(-1, check=True)
```
```suggestion
            @cython.cfunc
            @cython.exceptval(-1, check=True)
```

scoder commented on docs/src/userguide/language_basics.rst:
```diff
@@ -404,20 +634,53 @@ returns small results.
+
+        .. code-block:: python
+
+            @cfunc
+            @exceptval(check=True)
```
```suggestion
            @cython.cfunc
            @cython.exceptval(check=True)
```

0dminnimda commented on docs/src/userguide/language_basics.rst:
```diff
@@ -119,40 +202,75 @@ You can read more about them in :ref:`extension-types`.
-respectively, as statically typing variables with these Python
-types has zero advantages.
+
+.. Note:: The Python types ``int``, ``long``, and ``float`` are not available for static
```
```suggestion
.. note:: The Python types ``int``, ``long``, and ``float`` are not available for static
```

0dminnimda reviewed (changes requested):
Also we could use note in line 1122: "Note 1: Pyrex originally provided a function ..."

scoder commented:
Seems good enough for merging. Thanks. We can fix any remaining issues afterwards in separate PRs.

0dminnimda commented:
also @matusvalo you may want to check out [my package](https://pypi.org/project/translation-comparator/) that should help with pythonizing process
